### PR TITLE
gh-112087: Use QSBR technique for list_new/clear for free-thread build

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-112087

### CODE REVIEW

File: Lib/test/test_list.py [hunk 1]
```diff
@@ -230,6 +230,7 @@ def __eq__(self, other):
         self.assertFalse(list3 == list4)
 
     @cpython_only
+    @unittest.skipIf(Py_GIL_DISABLED, 'Only for the default build')
```
Comment (corona10) [inline Lib/test/test_list.py]: In free-thread build

```
  File "/Users/user/oss/cpython/Lib/test/test_list.py", line 237, in test_preallocation
    self.assertEqual(iter_size, sys.getsizeof(list([0] * 10)))
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 176 != 136
```

File: Objects/listobject.c [hunk 2]
```diff
@@ -498,11 +597,21 @@ static PyObject *
 list_item(PyObject *aa, Py_ssize_t i)
 {
     PyListObject *a = (PyListObject *)aa;
+    PyObject *item = NULL;
+    Py_BEGIN_CRITICAL_SECTION(a);
+#ifdef Py_GIL_DISABLED
+    if (!_PyObject_GC_IS_SHARED(a)) {
+        _PyObject_GC_SET_SHARED(a);
+    }
```
Comment (colesbury) [inline Objects/listobject.c]: I think the logic here should be if `a` is not owned by the current thread and not already shared.

File: Objects/listobject.c [hunk 3]
```diff
@@ -498,11 +597,21 @@ static PyObject *
 list_item(PyObject *aa, Py_ssize_t i)
 {
     PyListObject *a = (PyListObject *)aa;
+    PyObject *item = NULL;
+    Py_BEGIN_CRITICAL_SECTION(a);
```
Comment (colesbury) [inline Objects/listobject.c:506]: We will want the fast path to do this without locking soon.

File: Objects/listobject.c [hunk 4]
```diff
@@ -31,6 +31,141 @@ get_list_freelist(void)
...
+    return items;
+}
+
+/* Ensure ob_item has room for at least newsize elements, and set
+ * ob_size to newsize.  If newsize > ob_size on entry, the content
+ * of the new slots at exit is undefined heap trash; it's the caller's
+ * responsibility to overwrite them with sane values.
+ * The number of allocated elements may grow, shrink, or stay the same.
+ * Note that self->ob_item may change, and even if newsize is less
+ * than ob_size on entry.
```
Comment (colesbury) [inline Objects/listobject.c]: This comment doesn't seem accurate here

File: Objects/listobject.c [hunk 5]
```diff
@@ -970,7 +1093,7 @@ list_extend_fast(PyListObject *self, PyObject *iterable)
     // an overflow on any relevant platform.
     assert(m < PY_SSIZE_T_MAX - n);
     if (self->ob_item == NULL) {
-        if (list_preallocate_exact(self, n) < 0) {
+        if (list_ensure_capacity(self, m, n) < 0) {
```
Comment (colesbury) [inline Objects/listobject.c]: IIRC, the idea behind `list_ensure_capacity` was to subsume these extra checks. i.e. to just write:

```
if (list_ensure_capacity(self, m, n) < 0) {
  return NULL;
}
...
PyList_SET_SIZE(self, m + n);
```

No `self->ob_item == NULL` check or  extra `list_resize` call.

See https://github.com/colesbury/nogil-3.12/blob/cedde4f5ec3759ad723c89d44738776f362df564/Objects/listobject.c#L1206-L1220

File: Objects/listobject.c [hunk 6]
```diff
@@ -347,7 +446,7 @@ PyList_Insert(PyObject *op, Py_ssize_t where, PyObject *newitem)
 int
 _PyList_AppendTakeRefListResize(PyListObject *self, PyObject *newitem)
 {
-    Py_ssize_t len = Py_SIZE(self);
+    Py_ssize_t len = PyList_GET_SIZE(self);
```
Comment (colesbury) [inline Objects/listobject.c]: Why change to `PyList_GET_SIZE`? It doesn't look like it's necessary to me.

**Review by colesbury — COMMENTED** [review]
Thanks for taking this on, and sorry for the delay in reviewing this.

The `list_clear` change looks good, but I found the other parts difficult to review because the PR includes changes that aren't necessary for using QSBR. In particular:

- The `list_good_size` is a memory optimization so that requested sizes match mimalloc's internal size classes, but it's not necessary for thread-safety. Additionally, the code assumes that the allocated size is stored in the allocated `ob_item` array, but that's not implemented yet, so it's hard to tell if the code is correct.
- Similarly, `list_ensure_capacity` was written to simplify some resize checks given that we will generally want to match mimalloc's size classes rather than do the exact size allocations, but it's not necessary for correctness and not used here as it was used in the nogil-3.12 fork, which makes it harder to review.

I think it'll be easier to review (and maintain) to start with the narrow thread-safety changes and tackle the memory optimizations later.

File: Objects/listobject.c [hunk 7]
```diff
@@ -151,61 +193,18 @@ _PyList_DebugMallocStats(FILE *out)
...
-#endif
-    {
-        op = PyObject_GC_New(PyListObject, &PyList_Type);
-        if (op == NULL) {
-            return NULL;
+    PyListObject *op = list_new_prealloc(size);
+    if (op && op->ob_item) {
+        PyObject **items = op->ob_item;
+        for (Py_ssize_t i = 0, n = op->allocated; i < n; i++) {
+            FT_ATOMIC_STORE_PTR_RELEASE(items[i], NULL);
```
Comment (colesbury) [inline Objects/listobject.c]: I'm not sure we need to initialize `ob_item` here at all, but even if we do, we don't need atomic stores.

As written, `list_new_prealloc` already initializes the contents of `ob_item` to NULL. (But I'm not sure that we want `list_new_prealloc` to do that).

File: Objects/listobject.c [hunk 8]
```diff
@@ -31,6 +31,48 @@ get_list_freelist(void)
...
+            return NULL;
+        }
+    }
+    if (size <= 0) {
+        op->ob_item = NULL;
+        op->allocated = 0;
+    }
+    else {
+        size_t capacity = size;
+        PyObject **items = (PyObject **) PyMem_Calloc(size, sizeof(PyObject *));
```
Comment (colesbury) [inline Objects/listobject.c]: The previous code did not zero initialize `ob_item`. The callers to `list_new_prealloc` were responsible for initializing ob_item.

**Review by colesbury — COMMENTED** [review]
What is the motivation behind the refactoring of `list_new_prealloc` and `PyList_New()`? It doesn't appear to be necessary for thread-safety. I'm a bit concerned that the change to `list_new_prealloc` to zero initialize `ob_item` via `PyMem_Calloc` may have a performance impact on the default build.

**Comment by corona10** [comment]
> What is the motivation behind the refactoring of list_new_prealloc and PyList_New()? I

Ah, it was the legacy of my first patch, including mimalloc-based optimization.
I revert them all :)

**Review by colesbury — APPROVED** [review]

**Comment by corona10** [comment]
* test_sched: Already leaked but increased -> test_sched leaked [1, 0, 0] memory blocks, sum=1 (this is fine) to test_sched leaked [5, 4, 4] memory blocks, sum=13
* test.test_concurrent_futures.test_shutdown looks fine in my loacl.

```
Using random seed: 2687821224
Raised RLIMIT_NOFILE: 256 -> 1024
0:00:00 load avg: 2.41 Run 1 test sequentially
0:00:00 load avg: 2.41 [1/1] test_concurrent_futures.test_shutdown
beginning 6 repetitions. Showing number of leaks (. for 0 or less, X for 10 or more)
123:456
XX1 ...
test_concurrent_futures.test_shutdown passed in 1 min 13 sec

== Tests result: SUCCESS ==

1 test OK.

Total duration: 1 min 13 sec
Total tests: run=47 skipped=3
Total test files: run=1/1
Result: SUCCESS
```

**Comment by corona10** [comment]
Issue is triggered because https://github.com/python/cpython/blob/5dc8c84d397110f9edfa56793ad8887b1f176d79/Objects/obmalloc.c#L1021-L1023 it not executed.

@colesbury Is there a way to trigger _PyMem_ProcessDelayed by force?

**Comment by corona10** [comment]
With https://github.com/python/cpython/pull/116237

```
➜  cpython git:(gh-112087-leak-tmp) ✗ ./python.exe -m test test_sched -R 3:3
Using random seed: 2623432415
Raised RLIMIT_NOFILE: 256 -> 1024
0:00:00 load avg: 2.50 Run 1 test sequentially
0:00:00 load avg: 2.50 [1/1] test_sched
beginning 6 repetitions. Showing number of leaks (. for 0 or less, X for 10 or more)
123:456
XX4 5.4
test_sched leaked [0, 0, 1] references, sum=1 (this is fine)
test_sched leaked [5, -16, 4] memory blocks, sum=-7 (this is fine)

== Tests result: SUCCESS ==

1 test OK.

Total duration: 3.4 sec
Total tests: run=11
Total test files: run=1/1
Result: SUCCESS
```

**Comment by corona10** [comment]
Okay https://github.com/python/cpython/pull/116237 fix the issue, but I prefer https://github.com/python/cpython/pull/116238

**Pull request merged**